### PR TITLE
Parallel parsing data formats

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New feature

Changelog entry (up to few sentences, not needed for non-significant PRs):

(!) This feature enabled by default. (!)  
Parallel parsing is carried out thanks to ParallelParsingBlockInputStream class. There are 3 diffrent roles: Segmantator, Parser and Reader. Only Parser is multithreaded. So, how this class works. Segmentator cut the original file (or smth from ReadBuffer) into small pieces (you can control it with min_chunk_size_for_parallel_parsing setting). Then many parsers (also you can use max_threads_for_parallel_parsing setting for tuning) turn these pieces into Blocks. After that Blocks will be inserted into table without any shuffling (because it is order-preserving parallel parsing).

Old PR:
https://github.com/yandex/ClickHouse/pull/5372

### CODE REVIEW

**Comment by akuzm** [comment]
Do we really have to disable parallel parsings for so many tests? For several-line examples that are usually tested, parallel parsing doesn't make any difference. And I'm not sure what difference it makes for big files, so maybe we just have to leave it always on. 

If we do have to disable it for the tests, we can disable it globally in the configuration of the server that is used for the tests.

**Comment by alexey-milovidov** [comment]
> And I'm not sure what difference it makes for big files, so maybe we just have to leave it always on.

It will make the order of data non-deterministic.

The task №2 for @nikitamikhaylov is to make an option for order-preserving parallel parsing of data formats.

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                return;
+            }
+
+        }
+
+        current_unit.block = current_unit.reader->read();
+
+        {
+            std::lock_guard missing_values_lock(missing_values_mutex);
+            block_missing_values = current_unit.reader->getMissingValues();
```
Comment (nikitamikhaylov) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp]: Wrong placement

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.h [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        std::lock_guard missing_values_lock(missing_values_mutex);
+        return block_missing_values;
+    }
+
+private:
+
+    const std::atomic<size_t> max_threads_to_use;
+    const size_t min_chunk_size;
+
+    std::atomic<bool> is_exception_occured{false};
```
Comment (akuzm) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.h]: Is it possible to cancel when exception occurs, so that we don't need this variable?

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        {
+            std::unique_lock lock(mutex);
+            if (current_unit.is_last_unit || current_unit.readbuffer.position() == nullptr)
+            {
+                current_unit.block = Block();
+                current_unit.status = READY_TO_READ;
+                reader_condvar.notify_all();
+                return;
+            }
```
Comment (akuzm) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp]: Looks like it is not a last unit with data, but an empty past-the-end unit.
- should rename to something like `is_past_the_end`;
- do we need to spawn the parser thread for a past-the-end unit, maybe we can just do the related things from Segmentator thread?

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.h [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    ThreadPool pool;
+    // Reading and segmentating the file
+    ThreadFromGlobalPool segmentator_thread;
+
+    // Function to segment the file. Then "parsers" will parse that segments.
+    FormatFactory::FileSegmentationEngine file_segmentation_engine;
+
+    enum ProcessingUnitStatus
+    {
+        READY_TO_INSERT,
```
Comment (akuzm) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.h]: Is this actually an empty unit (i.e. has no file segment and no parsed block, and can be reused)?

File: dbms/src/Formats/FormatFactory.cpp [hunk 5]
```diff
@@ -100,6 +103,34 @@ BlockInputStreamPtr FormatFactory::getInput(
         return input_getter(buf, sample, context, max_block_size, callback ? callback : ReadCallback(), format_settings);
     }
 
+    const Settings & settings = context.getSettingsRef();
+    const auto & file_segmentation_engine = getCreators(name).file_segmentation_engine;
+
+    if (name != "Values" && settings.input_format_parallel_parsing && file_segmentation_engine)
```
Comment (akuzm) [inline dbms/src/Formats/FormatFactory.cpp]: `name != "Values"` -- don't forget to disable the Values engine properly.

File: dbms/src/Processors/Formats/Impl/CSVRowInputFormat.cpp [hunk 6]
```diff
@@ -433,4 +433,67 @@ void registerInputFormatProcessorCSV(FormatFactory & factory)
     }
 }
 
+void registerFileSegmentationEngineCSV(FormatFactory & factory)
+{
+    factory.registerFileSegmentationEngine("CSV", [](
```
Comment (akuzm) [inline dbms/src/Processors/Formats/Impl/CSVRowInputFormat.cpp]: Could you make these plain functions instead of lambdas? It's better for call stacks.

File: dbms/src/IO/ReadHelpers.h [hunk 7]
```diff
@@ -911,4 +912,10 @@ void skipToNextLineOrEOF(ReadBuffer & buf);
 /// Skip to next character after next unescaped \n. If no \n in stream, skip to end. Does not throw on invalid escape sequences.
 void skipToUnescapedNextLineOrEOF(ReadBuffer & buf);
 
+/** Return buffer eof() result.
+  * If there is no pending data in buffer or it was explicitly asked
+  *  save current state to memory.
+  */
+bool eofWithSavingBufferState(ReadBuffer & buf, DB::Memory<> & memory, size_t & used_size, char * & begin_pos, bool save_buffer_state = false);
```
Comment (akuzm) [inline dbms/src/IO/ReadHelpers.h]: My mind just goes blank when I see this function and how it's used. At the very least, it should have a sane name, and no bool flag that modifies behavior.

File: dbms/tests/queries/0_stateless/00395_nullable.sql [hunk 8]
```diff
@@ -102,11 +102,11 @@ SELECT col1 FROM test1_00395 ORDER BY col1 ASC;
 SELECT '----- Insert. Source and target columns have same types up to nullability. -----';
 DROP TABLE IF EXISTS test1_00395;
 CREATE TABLE test1_00395(col1 Nullable(UInt64), col2 UInt64) Engine=Memory;
-DROP TABLE IF EXISTS test2;
-CREATE TABLE test2(col1 UInt64, col2 Nullable(UInt64)) Engine=Memory;
+DROP TABLE IF EXISTS test2_00395;
```
Comment (akuzm) [inline dbms/tests/queries/0_stateless/00395_nullable.sql]: Changes in these three tests look unrelated to this PR.

File: dbms/src/IO/SharedReadBuffer.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+
+namespace DB
+{
+
+/** Allows many threads to read from single ReadBuffer
+  * Each SharedReadBuffer has his own memory,
+  *  which he filles under shared mutex using FileSegmentationEngine.
+  */
+class SharedReadBuffer : public BufferWithOwnMemory<ReadBuffer>
```
Comment (akuzm) [inline dbms/src/IO/SharedReadBuffer.h]: Looks like we don't use it anymore.

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.h (hunk 4 above)
Comment (nikitamikhaylov) [reply dbms/src/DataStreams/ParallelParsingBlockInputStream.h]: It means exactly that unit is ready to insert. It may contains previous junk file segment.

File: dbms/src/Formats/FormatFactory.h [hunk 10]
```diff
@@ -41,6 +42,16 @@ class FormatFactory final : private boost::noncopyable
     /// It's initial purpose was to extract payload for virtual columns from Kafka Consumer ReadBuffer.
     using ReadCallback = std::function<void()>;
 
+    /** Fast reading data from buffer and save result to memory.
+      * Reads at least min_chunk_size bytes and some more until the end of the chunk, depends on the format.
+      * Used in SharedReadBuffer.
```
Comment (akuzm) [inline dbms/src/Formats/FormatFactory.h]:
```suggestion
      * Used by ParallelParsingBlockInputStream.
```

File: dbms/src/Core/Settings.h [hunk 11]
```diff
@@ -108,6 +108,10 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingBool, distributed_group_by_no_merge, false, "Do not merge aggregation states from different servers for distributed query processing - in case it is for certain that there are different keys on different shards.") \
     M(SettingBool, optimize_skip_unused_shards, false, "Assumes that data is distributed by sharding_key. Optimization to skip unused shards if SELECT query filters by sharding_key.") \
     \
+    M(SettingBool, input_format_parallel_parsing, true, "Enable parallel parsing for several data formats (JSON, TSV, TKSV, Values, CSV).") \
+    M(SettingUInt64, max_threads_for_parallel_reading, 10, "The maximum number of threads to parallel reading. By default, it is set to max_threads.") \
+    M(SettingUInt64, min_chunk_size_for_parallel_reading, (1024 * 1024), "The minimum chunk size in bytes, which each thread tries to parse under mutex in parallel reading.") \
+    \
```
Comment (akuzm) [inline dbms/src/Core/Settings.h]: We should mention the new settings in the docs, and maybe add a couple of lines about how the parallel parsing works.

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                std::unique_lock lock(mutex);
+                current_unit.is_last_unit = true;
+                current_unit.status = READY_TO_PARSE;
+                scheduleParserThreadForUnitWithNumber(current_unit_number);
+                break;
+            }
+
+            std::unique_lock lock(mutex);
+            current_unit.status = READY_TO_PARSE;
+            scheduleParserThreadForUnitWithNumber(current_unit_number);
```
Comment (akuzm) [inline dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp]: Looks like we don't need the lock here -- schedule() call is a sufficient barrier to make the current_unit changes seen by the parser thread. A caveat: to conform to the C++ standard, current_unit.status should be std::atomic<>. A concurrent read may also occur from reader thread, and without atomic<> it would be UB.

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.h (hunk 2 above)
Comment (nikitamikhaylov) [reply dbms/src/DataStreams/ParallelParsingBlockInputStream.h]: It has brainy logic. Exception occurs in parser's threads and we have to signalize about it to reader's thread. If we used *is_cancel* variable, than _ThreadPool_ and other threads would never stop. (Because we use _CAS_ in _cancel()_).

File: dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp (hunk 3 above)
Comment (akuzm) [reply dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp]: Fixed here https://github.com/ClickHouse/ClickHouse/pull/7780
